### PR TITLE
Challenge2 read from file

### PR DESCRIPTION
**Pull request by kostana1**
code modified to read question, answers to them and get result. 

To be added later some basics validations

### CODE REVIEW

File: src/main/java/com/Utils/AppProperties.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public class AppProperties {
+
+    private  final String FILENAME = "application.properties";
+    private static final AppProperties properties_file = new AppProperties();
```
Comment (martojdk) [inline src/main/java/com/Utils/AppProperties.java]: constants should be uppercase letters , meaning **PROPERTIES_FILE**

File: src/main/java/com/Utils/AppProperties.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    private Properties properties = new Properties();
+
+    private AppProperties() {
```
Comment (martojdk) [inline src/main/java/com/Utils/AppProperties.java]: Great idea to have this kind of functionality in a separate class :) Very pleasantly surprised by the use of the singleton pattern(i hope you knew why are you doing it like this😄) 
I would only suggest separating the logic for populating the properties class in a separate method and also create the instance of the private fields in the constructor. Also maybe rename the class to something more meaningful

File: src/main/java/com/company/Main.java [hunk 3]
```diff
@@ -72,6 +76,28 @@ public static void printOptions() {
         System.out.println(PRINT_OPTIONS_MAIN_MENU);
     }
 
+    public static void createPersonWithAllAttributesToFile() {
+
+        CreatePersonService createPersonService = new CreatePersonService();
+        Person newPerson = new Person(createPersonService.createPersonName(), createPersonService.createPersonGender(), createPersonService.createPersonDateOfBirth(), createPersonService.createPersonInterests(), createPersonService.createPersonStatus());
+
+
+        try (FileWriter fileWriter = new FileWriter("personsList.txt", true);
```
Comment (martojdk) [inline src/main/java/com/company/Main.java]: This filename could be written in the application.properties

File: src/main/java/com/quiz/FindPersonByUuidAndAskQuestion.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    private static final String PERSON_UUID_BELONGS_TO = "This uuid belongs to %s\n";
+    public static final String ANSWER_NOW_QUESTION = "Answer now %s's question...\n";
+    public static final String LINE_SEPARATOR = "*******************";
+    public static final String RESULT = "Your result is %d";
+
+    private static PersonService personService = new PersonService();
+    private static Question question = new Question("test");
+    private static Scanner scanner = new Scanner(System.in);
+
+    public static void readFromFile() {
```
Comment (martojdk) [inline src/main/java/com/quiz/FindPersonByUuidAndAskQuestion.java]: 1.What do you read from file? Method naming is poor
2.This method does 2 things - 1.Populates some list 2.Asks some questions. Methods should have only one responsibility. So you can extract one of the two things in another method
3.Don't you think that populating the list of persons is a responsibility of the PersonService? I mean that this method should probably not be here
4.Why static?

File: src/main/java/com/quiz/FindPersonByUuidAndAskQuestion.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                EGender gender = EGender.valueOf(personData[2]);
+                Date date = CommonUtils.formatDateOfBirth(personData[3]);
+                String interest = personData[4];
+                EStatus status = EStatus.valueOf(personData[5]);
+
+                String questionLine;
+                questionLine = bufferedReader.readLine();
+                Question question = new Question(questionLine);
+                Person personFromFile = new Person(uuid, name, gender, date, interest, status, question);
+                personService.allPersons.add(personFromFile);
```
Comment (martojdk) [inline src/main/java/com/quiz/FindPersonByUuidAndAskQuestion.java]: Don't you have addPerson method? I don't think allpersons should be accessible

File: src/main/java/com/quiz/Question.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    private String questionDescription;
+    public List<Answer> questionAnswers;
+
+    public Question(String questionDescription) {
+        this.questionDescription = questionDescription;
+        this.questionAnswers = new ArrayList<>();
+    }
+
+    private Answer findAnswers(String answerDescription) {
```
Comment (martojdk) [inline src/main/java/com/quiz/Question.java]: This is a POJO(plain old java object) or model class, if you want functionalities create service for it.

File: src/main/java/com/service/CreatePersonService.java [hunk 7]
```diff
@@ -14,12 +15,18 @@
...
     public static final String INVALID_INPUT = "Invalid %s input";
-    public static final String SUCCESSFUL_INPUT = "%s successfully added";
+    public static final String SUCCESSFUL_INPUT = "%s successfully added\n";
     public static final String INTEREST_CHAR_LIMIT = "Your interests have been reduced to 250 characters";
     public static final String NAME_VALIDATION_REGEX = "[a-zA-z]{3,}";
     public static final String PATTERN_REGEXP_YEAR = "\\d{4}-[01]\\d-[0-3]\\d";
     public static final String USE_INTEGERS_ONLY = "Use integers as per description";
 
+    public static final String FIRST_QUESTION = "Select a number from 1 to 10 to assess your fairness when answering ";
+    public static final String SECOND_QUESTION = "How often do you drink? " + "\n1 - I don't drink; 2 - Not very often; 3 - Every other day; 4 - Every day";
```
Comment (martojdk) [inline src/main/java/com/service/CreatePersonService.java]: PLEASE dont use pluses

Comment (kostana1) [reply src/main/java/com/service/CreatePersonService.java]: Already know you don't prefer them. Just forgot to delete them when cleaning the code few days ago

File: src/main/java/com/Utils/AppProperties.java (hunk 2 above)
Comment (kostana1) [reply src/main/java/com/Utils/AppProperties.java]: Just got a reminder from a previous comments of yours when saying "put these try catch blocks in separate method :D"  and did a research for more clean approach

File: src/main/java/com/quiz/FindPersonByUuidAndAskQuestion.java (hunk 5 above)
Comment (kostana1) [reply src/main/java/com/quiz/FindPersonByUuidAndAskQuestion.java]: Silly me. Thanks for the hint

File: src/main/java/com/service/CreatePersonService.java (hunk 7 above)
Comment (martojdk) [reply src/main/java/com/service/CreatePersonService.java]: Double or triple check your code when submitting

File: src/main/java/com/Utils/CreatePropertiesFile.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package com.Utils;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public class CreatePropertiesFile {
```
Comment (martojdk) [inline src/main/java/com/Utils/CreatePropertiesFile.java]: think about what this class does what is responsible for and think about the name afterwards

File: src/main/java/com/company/Main.java [hunk 9]
```diff
@@ -72,9 +73,29 @@ public static void printOptions() {
         System.out.println(PRINT_OPTIONS_MAIN_MENU);
     }
 
-    public static void createPersonWithAllAttributes() {
+    public static void createPersonWithAllAttributesToFile() {
+
+        Person newPerson = new Person(createPersonService.createPersonName(), createPersonService.createPersonGender(), createPersonService.createPersonDateOfBirth(), createPersonService.createPersonInterests(), createPersonService.createPersonStatus());
+
+        String personListFilePath = CreatePropertiesFile.getInstance().getProperty("personListFilePath");
```
Comment (martojdk) [inline src/main/java/com/company/Main.java]: Property name should be a constant

File: src/main/java/com/quiz/Question.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
+package com.quiz;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Question implements IQuestion {
```
Comment (martojdk) [inline src/main/java/com/quiz/Question.java]: Thats not what I meant. Basically you should do the same as the personservice

File: src/main/java/com/Utils/CreatePropertiesFile.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    private  final String FILENAME = "application.properties";
+    private static final CreatePropertiesFile PROPERTY_FILE = new CreatePropertiesFile();
```
Comment (martojdk) [inline src/main/java/com/Utils/CreatePropertiesFile.java]: usually this field is named **instance**

File: src/main/java/com/quiz/Question.java (hunk 10 above)
Comment (kostana1) [reply src/main/java/com/quiz/Question.java]: Renamed from Question to AnswerService, simply because its operating with answers not with questions

File: src/main/java/com/Utils/CreatePropertiesFile.java (hunk 8 above)
Comment (kostana1) [reply src/main/java/com/Utils/CreatePropertiesFile.java]: ReadPropertyClass?

File: src/main/java/com/person/Person.java [hunk 12]
```diff
@@ -2,21 +2,23 @@
...
-    public static final String TO_STRING = "%s,%s,%s,%s,%s,%s";
+    public static final String TO_STRING = "%s,%s,%s,%s,%s,%s,%s";
 
     private UUID uuid;
     private String name;
     private Date dateOfBirth;
     private String interests;
     private EGender gender;
     private EStatus status;
+    private AnswerService answerService;
```
Comment (martojdk) [inline src/main/java/com/person/Person.java]: Why is this here

File: src/main/java/com/person/Person.java [hunk 13]
```diff
@@ -51,14 +63,18 @@ public UUID getUuid() {
...
     }
 
+    public AnswerService getQuestion() {
+        return answerService;
+    }
+
     @Override
     public String toString() {
-        return String.format(TO_STRING, uuid, name, gender, dateOfBirth, interests, status);
+        return String.format(TO_STRING, uuid, name, gender, dateOfBirth, interests, status, answerService);
```
Comment (martojdk) [inline src/main/java/com/person/Person.java]: What will be printed here

File: src/main/java/com/person/Person.java [hunk 14]
```diff
@@ -69,7 +85,7 @@ public boolean equals(Object obj) {
 
         if(obj instanceof Person) {
             Person objectPerson = (Person) obj;
-            return this.name.equals(objectPerson.getName()) && this.gender == objectPerson.getGender() && this.dateOfBirth == objectPerson.getDateOfBirth() && this.status == objectPerson.getStatus() && this.interests.equals(objectPerson.getInterests());
+            return this.name.equals(objectPerson.getName()) && this.gender == objectPerson.getGender() && this.dateOfBirth == objectPerson.getDateOfBirth() && this.status == objectPerson.getStatus() && this.interests.equals(objectPerson.getInterests()) && this.answerService.equals(objectPerson.getQuestion());
```
Comment (martojdk) [inline src/main/java/com/person/Person.java]: is answer service equal to objectperson's question ? What is the logic behind this?

File: src/main/java/com/quiz/AnswerService.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
+package com.quiz;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class AnswerService implements IAnswerService {
+
+    private String questionDescription;
```
Comment (martojdk) [inline src/main/java/com/quiz/AnswerService.java]: Why do you need this?

File: src/main/java/com/quiz/AnswerService.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    public List<Answer> questionAnswers;
```
Comment (martojdk) [inline src/main/java/com/quiz/AnswerService.java]: Whats the point of this?

File: src/main/java/com/service/CreatePersonService.java [hunk 17]
```diff
@@ -154,4 +187,72 @@ public EStatus createPersonStatus() {
             }
         }
     }
+
+    public void readPropertyFileAndExecuteReadFileLinesMethod() throws IOException {
```
Comment (martojdk) [inline src/main/java/com/service/CreatePersonService.java]: I use your method and I dont care how you do it I just want to know what is the final result. Rename properly
Also don't throw generic exceptions, create one specific for your purpose and wrap the generic one in yours.For example if you have exception named for example **ReadPersonException** 
try {
// something throwing IOException
} catch(IOException e){
throw new ReadPersonException("some message",e);
}

File: src/main/java/com/service/CreatePersonService.java [hunk 18]
```diff
@@ -154,4 +187,72 @@ public EStatus createPersonStatus() {
...
+        String testFilePath = CreatePropertiesFile.getInstance().getProperty("testFilePath");
+        BufferedReader bufferedReader = new BufferedReader(new FileReader(testFilePath));
+        readFileLines(bufferedReader);
+
+    }
+
+    public void readFileLines(BufferedReader bufferedReader) throws IOException {
```
Comment (martojdk) [inline src/main/java/com/service/CreatePersonService.java]: Does this have to be public? Also same story for the exceptions. When you are throwing exceptions you should think about the guy who is going to handle them

File: src/main/java/com/service/CreatePersonService.java [hunk 19]
```diff
@@ -154,4 +187,72 @@ public EStatus createPersonStatus() {
             }
         }
     }
+
+    public void readPropertyFileAndExecuteReadFileLinesMethod() throws IOException {
+        String testFilePath = CreatePropertiesFile.getInstance().getProperty("testFilePath");
+        BufferedReader bufferedReader = new BufferedReader(new FileReader(testFilePath));
```
Comment (martojdk) [inline src/main/java/com/service/CreatePersonService.java]: Where do you close it?

File: src/main/java/com/utils/CreatePropertiesFile.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
+package com.utils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public class CreatePropertiesFile {
+
+    private  final String PROPERTY_NAME = "application.properties";
```
Comment (martojdk) [inline src/main/java/com/utils/CreatePropertiesFile.java]: This is not property name

File: src/main/java/com/utils/CreatePropertiesFile.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
+package com.utils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public class CreatePropertiesFile {
```
Comment (martojdk) [inline src/main/java/com/utils/CreatePropertiesFile.java]: is this a method name or a class name what is it? Also if you "create" something where is the create method or create logic at least?

File: src/main/java/com/utils/CreatePropertiesFile.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    private static final CreatePropertiesFile INSTANCE = new CreatePropertiesFile();
+    private Properties properties = new Properties();
+
+
+    public CreatePropertiesFile() {
+        loadPropertiesFromFile(PROPERTY_NAME);
+    }
+
+    protected Properties loadPropertiesFromFile(String fileName) {
+        ClassLoader loader = Thread.currentThread().getContextClassLoader();
```
Comment (martojdk) [inline src/main/java/com/utils/CreatePropertiesFile.java]: Is this multi-threaded application?

File: src/test/java/com/service/CreatePersonServiceTest.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    private CreatePersonService classUnderTest;
+    private AnswerService answerService;
+
+    @BeforeEach
+    void setUp() {
+        classUnderTest = new CreatePersonService();
+        answerService = new AnswerService(QUESTION);
+        answerService.questionAnswers.add(new Answer("Johny", 10));
+        answerService.questionAnswers.add(new Answer("Peter", 20));
+        answerService.questionAnswers.add(new Answer("Kondio", 30));
```
Comment (martojdk) [inline src/test/java/com/service/CreatePersonServiceTest.java]: 😄 That's how tests are written!

File: src/test/java/com/service/CreatePersonServiceTest.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+        persons = new ArrayList<>();
+        persons.add(new Person(UUID.randomUUID(), "Kondio", FEMALE, new Date(), "bla blah blah", EStatus.IN_RELATIONSHIP, answerService));
+
+    }
+
+    @Test
+    void readFileLines() throws IOException {
```
Comment (martojdk) [inline src/test/java/com/service/CreatePersonServiceTest.java]: 1. Methods should be pubic
2.Use static imports for mockito
3.Use Mockito.verify() after the classUnderTest is executed
Here you can use it like 
**verify(bufferedReader,times(1)).readLine();**

File: src/test/java/com/utils/CreatePropertiesFileTest.java [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+    @BeforeEach
+    void setUp() {
+        this.classUnderTest = new CreatePropertiesFile();
+        properties = new Properties();
+        propertySource = "C:/Петко/udemy/java master class/Martin_Project/testResource/test.txt";
+    }
+
+    @Test
+    void whenLoadingProperties_thenExpectAllPropertiesToBeFilled() {
+        classUnderTest.loadPropertiesFromFile(FILENAME);
```
Comment (martojdk) [inline src/test/java/com/utils/CreatePropertiesFileTest.java]: Where is the verification that all properties are filled

File: src/test/java/com/utils/CreatePropertiesFileTest.java [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Test
+    void givenRealProperties_whenLoading_thenNotNullExpectToBeReturned() {
+        assertNotNull(properties);
```
Comment (martojdk) [inline src/test/java/com/utils/CreatePropertiesFileTest.java]: what is this test here

File: src/main/java/com/utils/CreatePropertiesFile.java [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+                properties.load(inputStream);
+            }
+        }catch (IOException e) {
+            e.printStackTrace();
+        }
+        return properties;
+    }
+
+    public String getPropertyName() {
+        return PROPERTY_NAME;
```
Comment (martojdk) [inline src/main/java/com/utils/CreatePropertiesFile.java]: Why do you need this method?

File: src/test/java/com/utils/CreatePropertiesFileTest.java [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Test
+    void givenRealName_whenGettingProperty_thenRightPropertyFileToBeExecuted() {
+        assertEquals(classUnderTest.getPropertyName(), FILENAME);
+    }
+
+    @Test
+    void givenPropertyKey_whenGettingProperty_thenRightPropertySourceToBeExecuted() {
```
Comment (martojdk) [inline src/test/java/com/utils/CreatePropertiesFileTest.java]: You can retrieve all sorts of properties from the file not only propertySource(by the way what is propertySource ? ). I think you don't fully understand what your method does and what it can be used for

File: src/test/java/com/utils/CreatePropertiesFileTest.java [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+        assertEquals(classUnderTest.getProperty("testFilePath"), propertySource);
+    }
+
+    @Test
+    void whenReadingNull_thenExpectToThrowNullPointerException(){
```
Comment (martojdk) [inline src/test/java/com/utils/CreatePropertiesFileTest.java]: Should you expect this? Nullpointer is a runtime exception. You should know when your code can throw it and try to avoid it with some checks(check if something is null for example)

File: src/test/java/com/utils/ApplicationPropertyFileExtractorTest.java [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+    @Test
+    public void whenFillingProperties_thenExpectAllPropertiesToBeFilled() {
+        classUnderTest.fillPropertiesFromFile(properties);
+        properties.setProperty("some key", "some value");
+        classUnderTest.getProperty("some key");
+        assertEquals("some value", properties.getProperty("some key"));
+    }
+
+    @Test
+    public void testPrivateConstructor() throws Exception{
```
Comment (martojdk) [inline src/test/java/com/utils/ApplicationPropertyFileExtractorTest.java:32]: No need for this but okay

File: src/main/java/com/quiz/IQuestion.java [hunk 31]
```diff
@@ -0,0 +1,9 @@
+package com.quiz;
+
+public interface IQuestion {
```
Comment (martojdk) [inline src/main/java/com/quiz/IQuestion.java:3]: If you want this IQuestion thing so desperately, you can name it IBaseQuestion for example and the child class to be something like PersonQuestion

**Review by martojdk — APPROVED** [review]
